proposals/fibers/Explainer.md

```diff
@@ -0,0 +1,874 @@
+
+### Fiber instructions
+
+#### `fiber.new` Create a new fiber
+
+The `fiber.new` instruction creates a new fiber. The instruction has a literal operand which is the index of a function of type `[[ref fiber  r*] t*]->[r*]`, together with corresponding values on the argument stack.
+
+The result of the `fiber.new` instruction is a `fiber` which is the identifier for the newly created fiber. The identity of the fiber is also the first argument to the fiber function itself&mdash;this allows fibers to know their own identity in a straightforward way.
+
+The return values that the fiber function returns represent the return values of the fiber. This is reflected in the type signatures of the `fiber` and the function.
+
+The new fiber is created in a suspended state: it must be the case that the function referenced has one or more `event` blocks within the top-level block of that function. When the fiber is resumed, one of those `event` blocks will be entered as a result of the `fiber.resume`&mdash;depending on the actual event used to resume the fiber.
+
+>If a function is referenced in a `fiber.new` instruction has no `event` blocks at the toplevel, then this function will cause a trap when attempting to resume the new fiber&mdash;because, without any `event` blocks, the fiber will not be able to respond to the resume.
+
+#### `fiber.spawn` Create and enter a new fiber
+
+The `fiber.spawn` instruction creates a new fiber and immediately enters it. It requires a literal function operand of type `[[fiber r*] t*] -> [r*]`, together with values corresponding to `t*` on the value stack.
+
+The `fiber.spawn` instruction is analogous to a `call` instruction, the effect of the instruction is to leave on the value stack the return values of the fiber's function. However, as with other fibers, the new fiber may suspend; in which case there should be appropriate `event` blocks in the instruction stream enclosing the `fiber.spawn` instruction.
+
+Notice that only the newly created fiber has access to the reference that identifies the new fiber. If it is desired that the parent has access to this the code generator would typically insert a `fiber.suspend` instruction at the start of the function and pass out the `fiber` in an appropriate event description.
+
+#### `fiber.suspend` Suspend an active fiber
+
+The `fiber.suspend` instruction takes a fiber as an argument and suspends the fiber. The identified fiber must either be the `active` fiber, or a resume ancestor of the active fiber. 
+
+The _root_ ancestor fiber does not have an explicit identifier; and so it may not be suspended.
+
+The fiber that is suspended is marked `suspended`, and the the immediate resume parent of that fiber becomes the active fiber.
```

Comment:
The primary requirement is that (a) you can only suspend active fibers and (b) you can only resume a fiber that was suspended. How you achieve that is a matter of implementation. One way is to have three states: active, suspended and inuse. The latter are fibers that are between other fibers on the 'fiber' stack.

Reply:
Right, but if a resume ancestor of the active fiber is suspended, I would expect that all of its resume dependents down to the active fiber would also be marked suspended to become eligible for resumption. Is that right?

Reply:
No. Only the fiber that actually was suspended can be resumed. The other fibers which are hanging off it may not be resumed.
When a suspended fiber is resumed, what actually happens is that the innermost fiber that was last running gets woken up. Because that is where the suspended fiber was running last.

Reply:
So that innermost fiber that gets resumed may not actually be an eligible target for the `fiber.resume` instruction? That seems extremely counterintuitive.

Reply:
I will make sense if you think about it :)